### ISSUE TITLE
Native image generation failed due too many loops.

### ISSUE THREAD

**Issue by lrxit** [oracle/graal#3000]
Native image generation failed due too many loops.

Source code is a grammar parser generated by ANTLR so there's no easy way to change number of loops due to the grammar itself.

Method contains 90 loops.

**ENVIRONMENT**

Windows 10 (x64) version 10.0.19041.630

GraalVM dev: 
#Fri Nov 20 03:10:07 GMT 2020
component_catalog=https\://raw.githubusercontent.com/graalvm/graalvm-ce-dev-builds/99fc14085a2d065760d3f54a1e5b7a833293481a/catalog.jdk_8.properties
OS_VERSION="5.2"
SOURCE=" jvmci\:0bbce9bf3339 compiler\:f73da1832213e19124ed4b8bf9d4e859019d19a9 graal-js\:4b91b7ee013092aeb1f76e3203653705d9205e0d graal-nodejs\:4b91b7ee013092aeb1f76e3203653705d9205e0d java-benchmarks\:f73da1832213e19124ed4b8bf9d4e859019d19a9 regex\:f73da1832213e19124ed4b8bf9d4e859019d19a9 sdk\:f73da1832213e19124ed4b8bf9d4e859019d19a9 substratevm\:f73da1832213e19124ed4b8bf9d4e859019d19a9 tools\:f73da1832213e19124ed4b8bf9d4e859019d19a9 truffle\:f73da1832213e19124ed4b8bf9d4e859019d19a9 vm\:f73da1832213e19124ed4b8bf9d4e859019d19a9"
OS_ARCH="amd64"
COMMIT_INFO={"compiler"\: {"commit.committer"\: "Martin Entlicher <martin.entlicher@oracle.com>", "commit.committer-ts"\: 1605826951, "commit.rev"\: "f73da1832213e19124ed4b8bf9d4e859019d19a9"}, "graal-js"\: {"commit.committer"\: "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts"\: 1603720197, "commit.rev"\: "4b91b7ee013092aeb1f76e3203653705d9205e0d"}, "graal-nodejs"\: {"commit.committer"\: "Gilles Duboscq <gilles.m.duboscq@oracle.com>", "commit.committer-ts"\: 1603720197, "commit.rev"\: "4b91b7ee013092aeb1f76e3203653705d9205e0d"}, "java-benchmarks"\: {"commit.committer"\: "Martin Entlicher <martin.entlicher@oracle.com>", "commit.committer-ts"\: 1605826951, "commit.rev"\: "f73da1832213e19124ed4b8bf9d4e859019d19a9"}, "regex"\: {"commit.committer"\: "Martin Entlicher <martin.entlicher@oracle.com>", "commit.committer-ts"\: 1605826951, "commit.rev"\: "f73da1832213e19124ed4b8bf9d4e859019d19a9"}, "sdk"\: {"commit.committer"\: "Martin Entlicher <martin.entlicher@oracle.com>", "commit.committer-ts"\: 1605826951, "commit.rev"\: "f73da1832213e19124ed4b8bf9d4e859019d19a9"}, "substratevm"\: {"commit.committer"\: "Martin Entlicher <martin.entlicher@oracle.com>", "commit.committer-ts"\: 1605826951, "commit.rev"\: "f73da1832213e19124ed4b8bf9d4e859019d19a9"}, "tools"\: {"commit.committer"\: "Martin Entlicher <martin.entlicher@oracle.com>", "commit.committer-ts"\: 1605826951, "commit.rev"\: "f73da1832213e19124ed4b8bf9d4e859019d19a9"}, "truffle"\: {"commit.committer"\: "Martin Entlicher <martin.entlicher@oracle.com>", "commit.committer-ts"\: 1605826951, "commit.rev"\: "f73da1832213e19124ed4b8bf9d4e859019d19a9"}, "vm"\: {"commit.committer"\: "Martin Entlicher <martin.entlicher@oracle.com>", "commit.committer-ts"\: 1605826951, "commit.rev"\: "f73da1832213e19124ed4b8bf9d4e859019d19a9"}}
GRAALVM_VERSION="21.0.0-dev"
OS_NAME="Windows"
JAVA_VERSION="1.8.0_272"

**STACK TRACE:**

Caused by: org.graalvm.compiler.core.common.PermanentBailoutException: Too many loops in method
        at org.graalvm.compiler.java.BciBlockMapping.makeLoopHeader(BciBlockMapping.java:1367)
        at org.graalvm.compiler.java.BciBlockMapping.computeBlockOrder(BciBlockMapping.java:1452)
        at org.graalvm.compiler.java.BciBlockMapping.computeBlockOrder(BciBlockMapping.java:1226)
        at org.graalvm.compiler.java.BciBlockMapping.build(BciBlockMapping.java:739)
        at org.graalvm.compiler.java.BciBlockMapping.create(BciBlockMapping.java:1561)
        at org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1008)
        at org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:986)
        at org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:84)
        at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:70)
        at org.graalvm.compiler.phases.Phase.run(Phase.java:49)
        at org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:212)
        at org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
        at org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
        at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:223)
        at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:357)
        at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:313)
        at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureParsed(MethodTypeFlow.java:302)
        at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:103)
        at com.oracle.graal.pointsto.DefaultAnalysisPolicy$DefaultSpecialInvokeTypeFlow.onObservedUpdate(DefaultAnalysisPolicy.java:368)
        at com.oracle.graal.pointsto.flow.TypeFlow.notifyObservers(TypeFlow.java:470)
        at com.oracle.graal.pointsto.flow.TypeFlow.update(TypeFlow.java:542)
        at com.oracle.graal.pointsto.flow.SourceTypeFlowBase.update(SourceTypeFlowBase.java:144)
        at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:547)
        at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:173)
        at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
        ... 4 more
Error: Image build request failed with exit status 1

**Comment by mcraj017** [comment]
@lrxit please provide the reproducer

**Comment by lrxit** [comment]
[here it is!](https://github.com/oracle/graal/files/5583674/loops.zip)

```
E:\graalvm-ce-java8-20.3.0\bin>native-image -jar "E:\issue\loops\target\loops-1.0-SNAPSHOT.jar" --no-fallback
[loops-1.0-SNAPSHOT:15260]    classlist:     696.81 ms,  1.15 GB
[loops-1.0-SNAPSHOT:15260]        (cap):   1,498.56 ms,  1.15 GB
[loops-1.0-SNAPSHOT:15260]        setup:   2,430.18 ms,  1.15 GB
[loops-1.0-SNAPSHOT:15260]     (clinit):      70.14 ms,  1.63 GB
[loops-1.0-SNAPSHOT:15260]   (typeflow):   3,061.06 ms,  1.63 GB
[loops-1.0-SNAPSHOT:15260]    (objects):   2,891.88 ms,  1.63 GB
[loops-1.0-SNAPSHOT:15260]   (features):      75.96 ms,  1.63 GB
[loops-1.0-SNAPSHOT:15260]     analysis:   6,186.02 ms,  1.63 GB
Error: Too many loops in method
Detailed message:
Call path from entry point to test.loops.NewMain.main(String[]):
        at test.loops.NewMain.main(NewMain.java:9)
        at com.oracle.svm.core.JavaMainWrapper.runCore(JavaMainWrapper.java:146)
        at com.oracle.svm.core.JavaMainWrapper.run(JavaMainWrapper.java:182)
        at com.oracle.svm.core.code.IsolateEnterStub.JavaMainWrapper_run_5087f5482cc9a6abc971913ece43acb471d2631b(generated:0)

Error: Use -H:+ReportExceptionStackTraces to print stacktrace of underlying exception
Error: Image build request failed with exit status 1
```

**Comment by lorenz** [comment]
Also ran into this with third-party code containing a function with 70 loops. It seems like the 64 limit is slightly too low, maybe 128 would be better.

**Comment by douglashg** [comment]
I also had the same error. Is there any forecast for correcting this error?

**Comment by mukel** [comment]
Fixed in 48c65d123ab4ec649b975e24048572f2a1e47b0c.
